### PR TITLE
Don't use wildcard imports in spack core

### PR DESCRIPTION
**Pull request by tgamblin**
This makes flake8 requirements stricter by:
1. Disallowing the use of wildcard imports in the Spack core
2. Restricting the use of wildcard imports to *only* `from spack import *` in packages.

This doesn't include `spack/__init__.py` yet, but the intent is to include it eventually.
Rationale: 
* not using wildcard imports allows us to detect undefined names and to find bugs, which we can't do if we use wildcards
* It also makes dependencies between spack modules more obvious, and will help us refactor the Spack core for an eventual PyPI package

Summary of this PR:
- [x] Add a `--all` option to `spack flake8` so that we can check all Python files
- [x] Separate the flake8 configuration between core and packages
- [x] Remove all wildcards from core and all *but* `from spack import *` from packages.
- [x] fix bugs discovered by the stricter flake8 settings.
- [x] additional small flake8 and bug fixes

I think this will make @jcftang happy.

### CODE REVIEW

**Comment by tgamblin** [comment]
@alalazo: can you look in particular at the last commit here?  It modifies a few parts of the module code.  I had to add fixes for 2 flake8 checks that apparently made their debut in the latest version of flake8 *while I was working on this*.

**Comment by adamjstewart** [comment]
And I thought I was crazy... This will be good though! It looks like it uncovered a lot of bugs, which is always good.

**Comment by tgamblin** [comment]
It's particularly good for finding bugs in error handling code where a variable name is either no longer in scope or has been changed.  And it helps for refactoring names.

I actually modified `spack flake8` to replace `from spack import *` with the names of all the variables in `spack` and in the module scope, but it didn't turn up any bugs, and it wasn't perfect -- there was a lot of noise from variables added by packages (`python`, `setup_py`, `luarocks`, etc.), and those would be a lot of trouble to check statically.

So I think that leaving  `from spack import *` in packages is worth the reduction in complexity for new contributors.

**Comment by healther** [comment]
Actually I'm not too sure that it reduces complexity for new contributors. ```import *``` doesn't give hints regarding what I, as a package creator, can make spack do. But having a list of functions that I can call on would give a starting point for searches in the spack source. So very much +1 for the core, but you might even think about replacing it in the ```package.py```'s too. It's generated code anyways, so it's not really a burden on anyone.

**Comment by adamjstewart** [comment]
I agree with @tgamblin on keeping `from spack import *` in packages. You shouldn't be searching for methods you can use via import commands, you should be reading through the Packaging Guide in the documentation.

File: lib/spack/spack/cmd/info.py [hunk 1]
```diff
@@ -189,12 +189,11 @@ def print_text_info(pkg):
...
-            (value.get('preferred', False), not key.isdevelop(), key)
-            for key, value in pkg.versions.items()
-        ]
-        l = sorted(l)
-        _, _, preferred = l.pop()
+        preferred = sorted(
+            pkg.versions, key=lambda v: (
+                pkg.versions[v].get('preferred', False),
+                not v.isdevelop(),
+                v)).pop()
```
Comment (alalazo) [inline lib/spack/spack/cmd/info.py]: Minor comment, but I find the previous version much more readable: the tuple is on a single line, sorting and popping are separate operations. If we want to rework it this way, maybe use:
```python
key_fn = lambda v: (pkg.versions[v].get('preferred', False), not v.isdevelop(), v)
preferred = sorted(pkg.versions, key=key_fn).pop()
```
?

File: lib/spack/spack/modules/lmod.py [hunk 2]
```diff
@@ -319,9 +318,10 @@ def unlocked_paths(self):
                 hierarchy = self.conf.hierarchy_tokens
                 available = self.conf.available
                 token2path = lambda x: self.token_to_path(x, available[x])
-                l = [x for x in hierarchy if x in item]
                 parts = []
-                for x in l:
+                for x in hierarchy:
```
Comment (alalazo) [inline lib/spack/spack/modules/lmod.py]: I find the lines below slightly less readable without a list comprehension.

File: lib/spack/spack/cmd/info.py (hunk 1 above)
Comment (tgamblin) [reply lib/spack/spack/cmd/info.py]: sure -- that's more readable!

File: lib/spack/spack/modules/lmod.py (hunk 2 above)
Comment (tgamblin) [reply lib/spack/spack/modules/lmod.py]: I guess I like the fact that my version only iterates through the data once.

File: lib/spack/spack/cmd/info.py (hunk 1 above)
Comment (tgamblin) [reply lib/spack/spack/cmd/info.py]: Personally I'm not such a fan of forbidding `l` as a variable name, but we already have exceptions to PEP8, and the point is to be as compliant as possible with Python OSS conventions, to make it easy to onboard people.  And in most cases I changed there were probably more descriptive names than `l` that could've been used so :man_shrugging:.

Comment (alalazo) [reply lib/spack/spack/cmd/info.py]: Completely agree with your statement above.

File: lib/spack/spack/modules/lmod.py (hunk 2 above)
Comment (alalazo) [reply lib/spack/spack/modules/lmod.py]: Optimizing is good, but I don't think people will use more than three / four items in the hierarchy :smile:

**Pull request merged**